addons/project/controllers/portal.py

```diff
@@ -94,40 +93,112 @@ def portal_my_project(self, project_id=None, access_token=None, **kw):
+            'user': request.env.user,
+            'project_accessible': project_accessible
         }
         return self._get_page_view_values(task, access_token, values, 'my_tasks_history', False, **kwargs)
 
-    @http.route(['/my/tasks', '/my/tasks/page/<int:page>'], type='http', auth="user", website=True)
-    def portal_my_tasks(self, page=1, date_begin=None, date_end=None, sortby=None, filterby=None, search=None, search_in='content', groupby=None, **kw):
-        values = self._prepare_portal_layout_values()
-        searchbar_sortings = {
+    def _task_get_searchbar_sortings(self):
+        return {
             'date': {'label': _('Newest'), 'order': 'create_date desc'},
             'name': {'label': _('Title'), 'order': 'name'},
             'stage': {'label': _('Stage'), 'order': 'stage_id, project_id'},
             'project': {'label': _('Project'), 'order': 'project_id, stage_id'},
             'update': {'label': _('Last Stage Update'), 'order': 'date_last_stage_update desc'},
+            'date_deadline': {'label': _('Deadline'), 'order': 'date_deadline asc'},
         }
-        searchbar_filters = {
-            'all': {'label': _('All'), 'domain': []},
+
+    def _task_get_searchbar_groupby(self):
+        values = {
+            'none': {'input': 'none', 'label': _('None'), 'order': 1},
+            'project': {'input': 'project', 'label': _('Project'), 'order': 2},
+            'stage': {'input': 'stage', 'label': _('Stage'), 'order': 3},
+            'user': {'input': 'user', 'label': _('Assigned to'), 'order': 6},
+            'customer': {'input': 'customer', 'label': _('Customer'), 'order': 7}
+        }
+        return dict(sorted(values.items(), key=lambda item: item[1]["order"]))
+
+    def _task_get_groupby_mapping(self):
+        return {
+            'project': 'project_id',
+            'stage': 'stage_id',
+            'user': 'user_id',
+            'customer': 'partner_id'
         }
-        searchbar_inputs = {
-            'content': {'input': 'content', 'label': Markup(_('Search <span class="nolabel"> (in Content)</span>'))},
-            'message': {'input': 'message', 'label': _('Search in Messages')},
-            'customer': {'input': 'customer', 'label': _('Search in Customer')},
-            'stage': {'input': 'stage', 'label': _('Search in Stages')},
-            'project': {'input': 'project', 'label': _('Search in Project')},
-            'all': {'input': 'all', 'label': _('Search in All')},
+
+    def _task_get_order(self, order, groupby):
+        if groupby == 'project':
+            order = "project_id, %s" % order  # force sort on project first to group by project in view
+        elif groupby == 'stage':
+            order = "stage_id, %s" % order  # force sort on stage first to group by stage in view
+        elif groupby == 'user':
+            order = "user_id, %s" % order
+        elif groupby == 'customer':
+            order = "partner_id, %s" % order
+        return order
```

Comment:
Since we have  the mapping of the groupby with this method rather than check the value of the `groupby` variable, why we would not use the dict returned in the `_task_get_groupby_mapping` method ? 🙂 
Like this:
```suggestion
        groupby_mapping = self._task_get_groupby_mapping()
        return '%s, %s' % (groupby_mapping[groupby], order)
```